text/3502-cargo-script.md

```diff
@@ -0,0 +1,1289 @@
+      will break unless the workspaces also owns the single-file package which
+      will break workflows for just creating a file anywhere to try out an
+      idea.
+- Cargo/rustc diagnostics and messages (including `cargo metadata`) should be
+  in terms of single-file packages and not any temporary files
+  - Easier to understand the messages
+  - Provides a higher quality of experience (doesn't feel like a hack or tacked on)
+  - Example implications:
+    - Most likely, we'll need single-file packages to be understood directly by
+      rustc so cargo doesn't have to split out the `.rs` content into a temp
+      file that gets passed to cargo which will cause errors to point to the
+      wrong file
+    - Most likely, we'll want to muck with the errors returned by `toml_edit`
+      so we render manifest errors based on the original source code which will require accurate span information.
+
+## Misc
+
+- Rejected: Defaulting to `RUST_BACKTRACE=1` for `cargo foo.rs` runs
+  - Enabling backtraces provides more context to problems, much like Python scripts, which helps with experiments
+  - This comes at the cost of making things worse for scripts
+  - Decided against it to minimize special casing
+  - See also [t-cargo zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Smarter.20.60RUST_BACKTRACE.60.20behavior.3F)
+- The package name is slugified according the stricter `cargo new` validation
+  rules, making it consistent across platforms as some names are invalid on
+  some platforms
+  - See [rust-lang/cargo#12255](https://github.com/rust-lang/cargo/pull/12255)
+
+## Command-line / interactive evaluation
+
+The [`cargo-script`](https://crates.io/crates/cargo-script) family of tools has a single command for
+- Run `.rs` files with embedded manifests
+- Evaluate command-line arguments (`--expr`, `--loop`)
+
+This behavior (minus embedded manifests) mirrors what you might expect from a
+scripting environment, minus a REPL.  We could design this with the future possibility of a REPL.
+
+However
+- The needs of `.rs` files and REPL / CLI args are different, e.g. where they get their dependency definitions
+- A REPL is a lot larger of a problem, needing to pull in a lot of interactive behavior that is unrelated to `.rs` files
+- A REPL for Rust is a lot more nebulous of a future possibility, making it pre-mature to design for it in mind
+
+Therefore, this RFC proposes we limit the scope of the new command to `cargo run` for single-file rust packages.
+
+## Naming
+[naming]: #naming
+
+Considerations:
+- The name should tie it back to `cargo` to convey that relationship
+- The command that is run in a `#!` line should not require arguments (e.g. not
+  `#!/usr/bin/env cargo <something>`) because it will fail.  `env` treats the
+  rest of the line as the bin name, spaces included.  You need to use `env -S`
+  but that wasn't supported on macOS at least, last I tested.
```

Comment:
A lot of decision is based around "`env -S` is not standard in particular macOS", but is it actually verified? @epage

1. On latest macOS Sonoma 14.4.1, `/usr/bin/env` *does* support an `-S` flag. [^1]
2. I have also got a macOS Big Sur 11.7.10 (2020) and its `env` also has the `-S` flag.
3. According to [Julia language's FAQ](https://docs.julialang.org/en/v1/manual/faq/#How-do-I-pass-options-to-julia-using-#!/usr/bin/env?), macOS supported `env -S` since Sierra 10.12 (2016), **and Sierra is the minimum supported macOS version [since Rust 1.74](https://blog.rust-lang.org/2023/09/25/Increasing-Apple-Version-Requirements.html)**.

Yes it is indeed still awkward to require the `-S` flag, but outright denying any alternatives that requires `-S` is not reasonable either.

[^1]: Additionally, unlike Linux, even skipping the `-S` flag it works just fine:
    ```
    #!/usr/bin/env cargo +nightly -Zscript
    fn main() { println!("works!"); }
    ```

Reply:
I've updated the text to not mention macOS.  I was going off of memory when I did that (it was a problem at a prior job and most problems came from macOS).  I have no idea if it was due to our Docker base images or something about macOS pre-Sierra.

I did confirm that `busybox` does not support `-S`.  The broader point is that its not portable.